src/scenarios/scenario_perception_pipeline.cpp

```diff
@@ -0,0 +1,156 @@
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Cihat Kurtuluş Altıparmak
+   Description: Benchmarking module to compare the effects of middlewares
+   against perception pipeline
+ */
+
+#include "moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp"
+
+namespace moveit
+{
+namespace middleware_benchmark
+{
+
+ScenarioPerceptionPipeline::ScenarioPerceptionPipeline(std::shared_ptr<MoveGroupInterface> move_group_interface_ptr)
+  : move_group_interface_ptr_(move_group_interface_ptr)
+{
+}
+
+std::tuple<int, int> ScenarioPerceptionPipeline::runTestCase(const nav_msgs::msg::Path& test_case)
+{
+  int success_number = 0;
+  int failure_number = 0;
+  for (auto& pose_stamped : test_case.poses)
+  {
+    bool is_successful = sendTargetPose(pose_stamped.pose);
+    if (is_successful)
+      success_number++;
+    else
+      failure_number++;
+  }
+
+  return { success_number, failure_number };
+}
+
+bool ScenarioPerceptionPipeline::sendTargetPose(const geometry_msgs::msg::Pose& target_pose)
+{
+  move_group_interface_ptr_->setPoseTarget(target_pose);
+
+  moveit::planning_interface::MoveGroupInterface::Plan plan_msg;
+  if (move_group_interface_ptr_->plan(plan_msg) == moveit::core::MoveItErrorCode::SUCCESS)
+  {
+    move_group_interface_ptr_->execute(plan_msg);
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+ScenarioPerceptionPipelineFixture::ScenarioPerceptionPipelineFixture()
+{
+  createTestCases();
+}
+
+void ScenarioPerceptionPipelineFixture::SetUp(::benchmark::State& /*state*/)
+{
+  if (node_.use_count() == 0)
+  {
+    node_ = std::make_shared<rclcpp::Node>("test_scenario_perception_pipeline_node",
+                                           rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true));
+
+    node_->get_parameter("selected_test_case_index", selected_test_case_index_);
+  }
+
+  if (move_group_interface_ptr_.use_count() == 0)
+  {
+    move_group_interface_ptr_ = std::make_shared<MoveGroupInterface>(node_, PLANNING_GROUP);
+  }
+}
+
+void ScenarioPerceptionPipelineFixture::TearDown(::benchmark::State& /*state*/)
+{
+}
+
+void ScenarioPerceptionPipelineFixture::createTestCases()
+{
+  readTestCasesFromFile(TEST_CASES_YAML_FILE);
+}
+
+nav_msgs::msg::Path ScenarioPerceptionPipelineFixture::selectTestCase(size_t test_case_index)
```

Comment:
In cases where the input is not changed you should make it const

Reply:
I fixed it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/1acebf40d7052970148dc6bd4a9ab84a2e5fd1bb commit. But I'm wondering  whether below code part should be returned const or not. Do you have any idea @sjahr ?

https://github.com/CihatAltiparmak/moveit_middleware_benchmark/blob/caa54960dd6febab1c82f7b63b93a38616339cda/src/scenarios/scenario_perception_pipeline.cpp#L131-L139

Reply:
I mark it as solved then.